### PR TITLE
Supports multiple MUCs and removing a MucClient.

### PR DESCRIPTION
**Pull request by bgrozev**

### CODE REVIEW

File: src/main/java/org/jitsi/xmpp/mucclient/MucClient.java [hunk 1]
```diff
@@ -506,4 +459,168 @@ private IQ handleIq(IQ iq)
...
+            throws SmackException.NotConnectedException,
+                   SmackException.NoResponseException,
+                   InterruptedException,
+                   XMPPException.XMPPErrorException,
+                   MultiUserChatException.MucAlreadyJoinedException,
+                   MultiUserChatException.NotAMucServiceException
+
+        {
+            // We're about to join or re-join the MUC.
+            lastPresenceSent = null;
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClient.java]: Yes. We want to use the IQ that smack generated for us, and add our extensions to it. If we re-join smack would have generated a new one.

File: src/main/java/org/jitsi/xmpp/mucclient/MucClient.java [hunk 2]
```diff
@@ -506,4 +459,168 @@ private IQ handleIq(IQ iq)
...
+                if (logger.isDebugEnabled())
+                {
+                    logger.info(
+                        "Not setting an extension yet, presence not sent.");
+                }
+                return;
+            }
+
+            // Remove the old extensions if present
+            extensions.forEach(
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClient.java]: These are the intended semantics, although I agree the name is confusing. It's consistent with the other methods, e.g. `setPresenceExtension` which does not clear all other extensions (but doesn't do a normal "add" either).

File: src/main/java/org/jitsi/xmpp/mucclient/MucClient.java [hunk 3]
```diff
@@ -506,4 +459,168 @@ private IQ handleIq(IQ iq)
...
+
+        /**
+         * Removes from our presence in the MUC any extensions with the given
+         * {@code elementName} and {@code namespace}.
+         * @param elementName the element name to match.
+         * @param namespace the namespace to match.
+         */
+        private void removePresenceExtension(String elementName, String namespace)
+        {
+            if (lastPresenceSent != null
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClient.java]: That's actually the reason I added `setPresenceExtensions`: it does not remove them one by one using this method, but iterates over the presence first and then sends a single stanza.

File: src/main/java/org/jitsi/xmpp/mucclient/MucClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbaldino) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClient.java]: ah weird.  ok.

File: src/main/java/org/jitsi/xmpp/mucclient/MucClientConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+     * @param prefix the common prefix (to be ignored) for the keys in the map.
+     * @param removeIncomplete whether to remove any incomplete (see
+     * {@link MucClientConfiguration#isComplete()}) entries from the returned
+     * collection, or to return all of them regardless.
+     *
+     * @return a list of {@link MucClientConfiguration}s described by the
+     * entries of {@code properties} with a prefix of {@code prefix}.
+     */
+    public static Collection<MucClientConfiguration> loadFromMap(
+        Map<String, String> properties, String prefix, boolean removeIncomplete)
```
[the remark this answers is not in the document]
Comment (bgrozev) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClientConfiguration.java]: I would have to add code to specifically read the required properties in more than one place (the code which reads from a ConfigurationService and the code which reads JSON)

File: src/main/java/org/jitsi/xmpp/mucclient/MucClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbaldino) [reply src/main/java/org/jitsi/xmpp/mucclient/MucClient.java]: oops, i misread. 👍

**Pull request merged**